pkg/services/oauthtoken/oauth_token_test.go

```diff
@@ -270,3 +150,458 @@ func (f *FakeAuthInfoStore) UpdateAuthInfo(ctx context.Context, cmd *login.Updat
+			},
+			expectedTokenRefreshFlag: false,
+			expectedTokenDuration:    maxOAuthTokenCacheTTL,
+		},
+		{
+			name: "should flag token refresh with id token is expired",
+			usr: &login.UserAuth{
+				OAuthIdToken: EXPIRED_JWT,
+			},
+			expectedTokenRefreshFlag: true,
+			expectedTokenDuration:    time.Second,
+		},
+		{
+			name: "should flag token refresh when expiry date is zero",
+			usr: &login.UserAuth{
+				OAuthExpiry: time.Unix(0, 0),
+			},
+			expectedTokenRefreshFlag: true,
+			expectedTokenDuration:    time.Second,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			token, needsTokenRefresh, tokenDuration := needTokenRefresh(tt.usr)
+
+			assert.NotNil(t, token)
+			assert.Equal(t, tt.expectedTokenRefreshFlag, needsTokenRefresh)
+			assert.Equal(t, tt.expectedTokenDuration, tokenDuration)
+		})
+	}
+}
+
+func TestOAuthTokenSync_tryGetOrRefreshOAuthToken(t *testing.T) {
+	timeNow := time.Now()
+	token := &oauth2.Token{
+		AccessToken:  "oauth_access_token",
+		RefreshToken: "refresh_token_found",
+		Expiry:       timeNow,
+		TokenType:    "Bearer",
+	}
+	type environment struct {
+		authInfoService *authinfotest.FakeService
+		cache           *localcache.CacheService
+		socialConnector *socialtest.MockSocialConnector
+		socialService   *socialtest.FakeSocialService
+
+		service *Service
+	}
+	tests := []struct {
+		desc          string
+		expectedErr   error
+		expectedToken *oauth2.Token
+		usr           *login.UserAuth
+		setup         func(env *environment)
+	}{
+		{
+			desc: "should refresh token if the token is expired",
+			usr: &login.UserAuth{
+				UserId:           int64(1234),
+				OAuthAccessToken: "new_access_token",
+				OAuthExpiry:      timeNow,
+			},
+			setup: func(env *environment) {
+				env.cache.Set("token-check-1234", token, 1*time.Minute)
+			},
+			expectedToken: &oauth2.Token{
+				AccessToken: "new_access_token",
+				Expiry:      timeNow,
+			},
+		},
+		{
+			desc: "should return ErrNotAnOAuthProvider error when the user is not an oauth provider",
+			usr: &login.UserAuth{
+				UserId:     int64(1234),
+				AuthModule: login.SAMLAuthModule,
+			},
+			expectedErr: ErrNotAnOAuthProvider,
+		},
+		{
+			desc: "should return ErrNoRefreshTokenFound error when the no refresh token was found",
+			usr: &login.UserAuth{
+				UserId:     int64(1234),
+				AuthModule: login.GenericOAuthModule,
+			},
+			expectedErr: ErrNoRefreshTokenFound,
+		},
+		{
+			desc: "should not refresh token if the token is not expired",
+			usr: &login.UserAuth{
+				UserId:            int64(1234),
+				AuthModule:        login.GenericOAuthModule,
+				OAuthAccessToken:  "oauth_access_token",
+				OAuthRefreshToken: "refresh_token_found",
+				OAuthExpiry:       timeNow,
+			},
+			expectedToken: token,
+			setup: func(env *environment) {
+				env.socialConnector.On("TokenSource", mock.Anything, mock.Anything).Return(oauth2.StaticTokenSource(token)).Once()
+			},
```

Comment:
This test does not do what it says.
```suggestion
			usr: &login.UserAuth{
				UserId:            int64(1234),
				AuthModule:        login.GenericOAuthModule,
				OAuthAccessToken:  token.AccessToken,
				OAuthRefreshToken: token.RefreshToken,
				OAuthExpiry:       timeNow.Add(time.Hour),
				OAuthTokenType:    "Bearer",
			},
			expectedToken: &oauth2.Token{
				AccessToken:  token.AccessToken,
				RefreshToken: token.RefreshToken,
				Expiry:       timeNow.Add(time.Hour),
				TokenType:    "Bearer",
			},
```